packages/cozy-realtime/src/CozyRealtime.js

```diff
@@ -1,360 +1,534 @@
-      return new Promise(resolve => {
-        this._socket.once('open', async () => {
-          await this.subscribe(eventName, type, handlerOrId, handlerOrUndefined)
-          resolve()
-        })
-      })
-    }
+   * When a websocket is closed by something else - reconnect
+   * @private
+   */
+  onWebSocketClose(event) {
+    logger.info('The current websocket was closed by a third party', event)
+    this.retryManager.onFailure(event)
+    this.reconnect()
+  }
 
-    return new Promise(resolve => {
-      const key = generateKey(eventName, type, id)
-      this.on(key, handler)
+  /* * * * * * * * * * * * * */
+  /* * Cozy Client EVENTS  * */
+  /* * * * * * * * * * * * * */
 
-      this._socket.once(`subscribe_${type}_${id}`, resolve)
-      this._socket.subscribe(type, id)
-    })
+  /**
+   * Register event handlers for cozy-client events
+   * @private
+   */
+  subscribeClientEvents() {
+    this.client.on('login', this.onClientLogin)
+    this.client.on('tokenRefreshed', this.onClientTokenRefreshed)
+    this.client.on('logout', this.onClientLogout)
   }
 
   /**
-   * Remove the given handler from the list of handlers for given
-   * doctype/document and event.
-   *
-   * @param {String}  type      Document doctype to unsubscribe from
-   * @param {String}  id        Document id to unsubscribe from
-   * @param {String}  eventName Event to unsubscribe from
-   * @param {Function}  handler   Function to call when an event of the
-   * given type on the given doctype or document is received from stack.
+   * Unregister event handlers for cozy-client events
+   * @private
    */
-  unsubscribe(eventName, type, handlerOrId, handlerOrUndefined) {
-    const { handler, id } = getHandlerAndId(handlerOrId, handlerOrUndefined)
-    validateParameters(eventName, type, id, handler)
-    const key = generateKey(eventName, type, id)
-
-    this.removeListener(key, handler)
+  unsubscribeClientEvents() {
+    this.client.removeListener('login', this.onClientLogin)
+    this.client.removeListener('tokenRefreshed', this.onClientTokenRefreshed)
+    this.client.removeListener('logout', this.onClientLogout)
+  }
 
-    if (!this._haveEventHandler()) {
-      this._resetSocket()
+  /**
+   * When the cozy-client instance successfully login - re-authenticate
+   * @private
+   */
+  onClientLogin() {
+    logger.info('Received login from cozy-client')
+    if (this.isWebSocketOpen()) {
+      this.authenticate()
+      // send subscriptions again, permissions may have changed
+      this.sendSubscriptions()
     }
   }
 
   /**
-   * Unsubscribe all handlers and close socket
+   * When the cozy-client instance refresh its access token - re-authenticate
+   * @private
    */
-  unsubscribeAll() {
-    this._getEventKeys()
-      .map(key => this._events[key].length > 0 && key)
-      .filter(Boolean)
-      .forEach(key => (this._events[key] = []))
+  onClientTokenRefreshed() {
+    // no need to authenticate again
+  }
```

Comment:
I understand the theory behind that case. But what if we start an Oauth flow, with the token that will expire in a few seconds, and then connect to a realtime? We will send an expired token, the client will refresh the token, and then we will have to wait the retry mechanism?

Reply:
We won't even have a retry mechanism in this case (as we won't be able torecover automatically anyways for a bad token).

And yes, you're right, we may have a very bad timing. I will add a reconnect there.